src/main/c/Objects/pyjmap.c

```diff
@@ -223,6 +218,153 @@ static PyObject* pyjmap_getiter(PyObject* obj)
     return result;
 }
 
+static PyObject* pyjmap_keys(PyObject* self, PyObject* args)
+{
+    jobject    keyset = NULL;
+    PyObject  *result = NULL;
+    PyJObject *pyjob  = (PyJObject*) self;
+    JNIEnv    *env    = pyembed_get_env();
+
+    if ((*env)->PushLocalFrame(env, JLOCAL_REFS) != 0) {
+        process_java_exception(env);
+        return NULL;
+    }
+
+    keyset = java_util_Map_keySet(env, pyjob->object);
+    if (process_java_exception(env) || !keyset) {
+        goto FINALLY;
+    }
+
+    result = jobject_As_PyObject(env, keyset);
+FINALLY:
+    (*env)->PopLocalFrame(env, NULL);
+    return result;
+}
+
+static PyObject* pyjmap_items(PyObject* self, PyObject* args)
+{
+    jobject    entrySet = NULL;
+    jobject    itr      = NULL;
+    PyObject  *pylist   = NULL;
+    PyObject  *result   = NULL;
+    int        size     = 0;
+    int        index    = 0;
+    PyJObject *pyjob    = (PyJObject*) self;
+    JNIEnv    *env      = pyembed_get_env();
+
+    if ((*env)->PushLocalFrame(env, JLOCAL_REFS) != 0) {
+        process_java_exception(env);
+    }
+
+    entrySet = java_util_Map_entrySet(env, pyjob->object);
+    if (process_java_exception(env) || !entrySet) {
```

Comment:
Done

Reply:
if process_java_exception() sets the python error(returns true) then we should not call PyErr_SetString because there can only be one python exception at a time. To set an error on null I think the code has to be structured like the error handling for next below(285-292) with 2 separate ifs. Also there are other places in this file where you check for null after process_java_exception and those would need similar updates.

Reply:
Done